### PR TITLE
Replace `failure` crate with `anyhow` and `thiserror`

### PR DESCRIPTION
**Pull request by tizoc**
`failure` was deprecated a while ago: https://github.com/rust-lang-nursery/failure/pull/347

This PR is a very mechanical (lots of find+replace) replacement of `failure` with [thiserror](https://github.com/dtolnay/thiserror) and [anyhow](https://github.com/dtolnay/anyhow) while leaving things very close to what they are ~(still some `TODO - NEWERRORS`s in a few cases that make use of more advanced features, like in `tezos/encoding`, so they need to be revised).~

- ~One here: https://github.com/tezedge/tezedge/blob/af902c3c8f084b42a15d75d88aa229da21c215d4/rpc/src/server/shell_handler.rs#L649~
- ~Three in this file: https://github.com/tezedge/tezedge/blob/af902c3c8f084b42a15d75d88aa229da21c215d4/tezos/encoding/src/binary_writer.rs~
- ~And this file in general: https://github.com/tezedge/tezedge/blob/af902c3c8f084b42a15d75d88aa229da21c215d4/tezos/encoding/src/error_context.rs~

It all works very similarly to before, but with one extra (and IMO important) advantage, the `#[from]` annotation that can be used to derive `impl From<OtherError>`, something we do a lot manually. I didn't update that in this PR, but it is something that we can start taking advantage of (except in this case that serves as an example: https://github.com/tezedge/tezedge/pull/605/files#diff-1e99e30dca4f07ae8dcf830c43aa3528cae270cc44b338dc86b01219f509bcc4L33 )

Before:

```rust
#[derive(Debug, Fail, PartialEq)]
pub enum PublicKeyError {
    #[fail(display = "Error constructing hash: {}", _0)]
    HashError(FromBytesError),
    #[fail(display = "Blake2b digest error: {}", _0)]
    Blake2bError(Blake2bError),
}

impl From<FromBytesError> for PublicKeyError {
    fn from(source: FromBytesError) -> Self {
        PublicKeyError::HashError(source)
    }
}

impl From<Blake2bError> for PublicKeyError {
    fn from(source: Blake2bError) -> Self {
        PublicKeyError::Blake2bError(source)
    }
}
```

After:

```rust
#[derive(Debug, Error, PartialEq)]
pub enum PublicKeyError {
    #[error("Error constructing hash")]
    HashError(#[from] FromBytesError),
    #[error("Blake2b digest error")]
    Blake2bError(#[from] Blake2bError),
}
```

thoughts?

### CODE REVIEW

**Comment by tizoc** [comment]
Another advantage of `#[from]` is that with it, the inner error is displayed automatically as part of the `Display` version of the wrapping error, there is no need to manually include that in the format string.

**Comment by tizoc** [comment]
@akoptelov for the `TODO - NEWERRORS`s in the `tezos/encoding` you are probably the best person to ask, I think you wrote the original code there.

**Comment by tizoc** [comment]
@bkontur @akoptelov just updated this to be in sync with current `develop`.

Still have to verify/update the code in `tezos/encoding`, and also I guess `tezedge/tezedge-ci-builder` needs to be updated to use a more recent nightly rust too.

**Comment by tizoc** [comment]
Oh and obviously, update code to take advantage of `#[from]` to automate `From` definitions and the part of the message that prints the inner errors. But that can be done after merged by everyone who touches related code.

**Comment by bkontur** [comment]
@tizoc @akoptelov 
there is a PR cooming soon to develop where I removed lots of usages of `failure` from rpc module:
https://github.com/tezedge/tezedge/pull/721/files

**Comment by tizoc** [comment]
@bkontur I would like to get rid of as many of the generic errors as possible (when I updated this PR today I noticed that many of the generic ones were gone). Adopting the changes in this PR will help with that too because it requires far less code to define errors that get chained from other inner errors.

**Comment by tizoc** [comment]
@akoptelov ok, just updated this now on top of your new encoder, should all be clean now. I have to revise it once more and verify that nothing is failing and then it can be merged.

**Review by akoptelov — COMMENTED** [review]
Please take a look at my comment about version.

File: apps/node_monitoring/Cargo.toml [hunk 1]
```diff
@@ -9,7 +9,7 @@ default-run = "node-monitoring"
 cfg-if = "1.0"
 clap = "2.33"
 chrono = "0.4"
-failure = "0.1"
+thiserror = "1"
```
[the remark this answers is not in the document]
Comment (tizoc) [reply apps/node_monitoring/Cargo.toml:12]: Ah yes, good catch.

**Pull request merged**